feat(ArangoDB): Add basic support for ArangoDB

enchantednatures opened the pull request:
This adds basic support for ArangoDB, a multi-modal Graph Database. This initial implementation phase supports primarily the dedicated query language: AQL and the json responses.

MattiasMTS commented on dbee/adapters/arango_driver.go:
```diff
@@ -0,0 +1,233 @@
+}
+
+func (a *arangoDriver) ListDatabases() (current string, available []string, err error) {
+	if a == nil || a.c == nil {
```
This goes for all the other method checks in this file.

enchantednatures replied:
I'm not that familiar with golang. I was unsure if there was risk of a nil receiver here. I will remove.

MattiasMTS commented on dbee/adapters/arango_driver.go:
```diff
@@ -0,0 +1,233 @@
+	}
+
+	log.Printf("Fetching columns for table: %s", opts.Table)
+
+	db, err := a.c.GetDatabase(context.Background(), a.dbName, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get database: %w", err)
+	}
+
+	aql := `
+	FOR n IN @@col
+		LIMIT 1000
+		FOR a IN ATTRIBUTES(n)
+		COLLECT attribute = a WITH COUNT INTO len
+		SORT len DESC
+		LIMIT 10
+		sort attribute
+		RETURN {attribute}`
+
+	bindVars := map[string]interface{}{"@col": opts.Table}
+	cursor, err := db.Query(context.Background(), aql, &arangodb.QueryOptions{BindVars: bindVars})
+	if err != nil {
+		return nil, fmt.Errorf("failed to execute query: %w", err)
+	}
+	defer cursor.Close()
+
+	columns := []*core.Column{}
+	var doc map[string]any
+
+	for cursor.HasMore() {
+		_, err := cursor.ReadDocument(context.Background(), &doc)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read document: %w", err)
+		}
+		column := fmt.Sprintf("%s", doc["attribute"])
+		columns = append(columns, &core.Column{Type: "collection", Name: column})
+	}
+
+	return columns, nil
+}
```
```suggestion
func (a *arangoDriver) Columns(opts *core.TableOptions) ([]*core.Column, error) {
	db, err := a.c.GetDatabase(context.Background(), a.dbName, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to get database: %w", err)
	}

	aql := `
	FOR n IN @@col
		LIMIT 1000
		FOR a IN ATTRIBUTES(n)
		COLLECT attribute = a WITH COUNT INTO len
		SORT len DESC
		LIMIT 10
		sort attribute
		RETURN {attribute}`

	bindVars := map[string]any{"@col": opts.Table}
	cursor, err := db.Query(context.Background(), aql, &arangodb.QueryOptions{BindVars: bindVars})
	if err != nil {
		return nil, fmt.Errorf("failed to execute query: %w", err)
	}
	defer cursor.Close()

	columns := make([]*core.Column, 0)
	doc := make(map[string]any)

	for cursor.HasMore() {
		if _, err := cursor.ReadDocument(context.Background(), &doc); err != nil {
			return nil, fmt.Errorf("failed to read document: %w", err)
		}
		column, ok := doc["attribute"].(string)
		if !ok {
			column = ""
		}
		columns = append(columns, &core.Column{Type: "collection", Name: column})
	}

	return columns, nil
}
```

MattiasMTS replied:
or return error if the type casting of "attribute" fails.

MattiasMTS commented on dbee/adapters/arango_driver.go:
```diff
@@ -0,0 +1,233 @@
+	db, err := a.c.GetDatabase(ctx, a.dbName, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get database: %w", err)
+	}
+
+	cursor, err := db.Query(ctx, query, nil)
+	if err != nil {
+		return nil, fmt.Errorf("query execution failed: %w", err)
+	}
+	defer cursor.Close()
+	next, hasNext := builders.NextNil()
+
+	next, hasNext = builders.NextYield(func(yield func(...any)) error {
+		if !cursor.HasMore() {
+			next, hasNext = builders.NextNil()
+		}
+
+		for {
+			if !cursor.HasMore() {
+				break
+			}
+			var doc interface{}
+
+			_, err = cursor.ReadDocument(ctx, &doc)
+			if err != nil {
+				return err
+			}
+
+			yield(NewArangoResponse(doc))
+		}
+
+		return nil
+	})
+
+	return builders.NewResultStreamBuilder().
+		WithNextFunc(next, hasNext).
+		WithHeader(core.Header{"Results"}).
+		WithMeta(&core.Meta{SchemaType: core.SchemaLess}).
+		Build(), nil
+}
```
```suggestion
func (a *arangoDriver) Query(ctx context.Context, query string) (core.ResultStream, error) {
	db, err := a.c.GetDatabase(ctx, a.dbName, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to get database: %w", err)
	}

	cursor, err := db.Query(ctx, query, nil)
	if err != nil {
		return nil, fmt.Errorf("query execution failed: %w", err)
	}
	defer cursor.Close()

	next, hasNext := builders.NextYield(func(yield func(...any)) error {
		for cursor.HasMore() {
			var doc any

			if _, err := cursor.ReadDocument(ctx, &doc); err != nil {
				return err
			}

			yield(NewArangoResponse(doc))
		}

		return nil
	})

	return builders.NewResultStreamBuilder().
		WithNextFunc(next, hasNext).
		WithHeader(core.Header{"Results"}).
		WithMeta(&core.Meta{SchemaType: core.SchemaLess}).
		Build(), nil
}
```

MattiasMTS commented on dbee/adapters/arango_driver.go:
```diff
@@ -0,0 +1,233 @@
+	return w.Bytes(), err
+}
+
+func (ar *arangoResponse) GobDecode(buf []byte) error {
+	var err error
+	r := bytes.NewBuffer(buf)
+	decoder := gob.NewDecoder(r)
+	err = decoder.Decode(&ar.Value)
+	if err != nil {
+		return err
+	}
+	return err
+}
```
```suggestion
func (ar *arangoResponse) GobDecode(buf []byte) error {
	r := bytes.NewBuffer(buf)
	decoder := gob.NewDecoder(r)
	return decoder.Decode(&ar.Value)
}
```

MattiasMTS commented on dbee/adapters/arango_driver.go:
```diff
@@ -0,0 +1,233 @@
+	return json.Marshal(ar.Value)
+}
+
+func (ar *arangoResponse) GobEncode() ([]byte, error) {
+	var err error
+	w := new(bytes.Buffer)
+	encoder := gob.NewEncoder(w)
+	err = encoder.Encode(ar.Value)
+	if err != nil {
+		return nil, err
+	}
+	return w.Bytes(), err
+}
```
```suggestion
func (ar *arangoResponse) GobEncode() ([]byte, error) {
	w := new(bytes.Buffer)
	encoder := gob.NewEncoder(w)
	if err := encoder.Encode(ar.Value); err != nil {
		return nil, err
	}
	return w.Bytes(), nil
}
```

MattiasMTS commented on dbee/adapters/arango_driver.go:
```diff
@@ -0,0 +1,233 @@
+		return nil, fmt.Errorf("failed to list databases: %w", err)
+	}
+
+	structures := make([]*core.Structure, len(databases))
```
nice clean! ✨

MattiasMTS reviewed (commented):
Thanks for your contribution @enchantednatures! 😄 ✨  much appreciated, I learned a new database today 😋 

Left some early comments on the adapter code - I'll take a second look on the integration tests. Very nice and appreciated the addition of integration test in the PR 👍

enchantednatures commented on dbee/adapters/arango.go:
```diff
@@ -0,0 +1,64 @@
+	if u.User != nil {
+		// Basic Authentication
+		username := u.User.Username()
+		password, _ := u.User.Password()
```
It is possible, albeit not recommended, to run Arango and not require a root password. I will add a test case to cover this scenario.

enchantednatures commented:
@MattiasMTS  thanks for taking the time to review! I've added a few changes, and applied your suggestions. Please take a look specifically at the latest changes in `Connect` I've added parsing options from the connection string query parameters.
